### PR TITLE
bitcoind v19 new RPC calls and tests

### PR DESCRIPTION
**Pull request by benthecarman**
First PR, this just adds the new RPC methods and tests for them. I am not sure how exhaustive the tests should be so I just put pretty basic ones in.

New RPCs

- `getblockfilter`

- `getbalances`

- `setwalletflag`

### CODE REVIEW

**Comment by nkohen** [comment]
Would you mind listing the RPCs implemented in the description?

**Comment by benthecarman** [comment]
> Would you mind listing the RPCs implemented in the description?

Added

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/V19BlockFilterRpc.scala [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+/**
+  * Gets the BIP158 filter for the specified block.
+  * This RPC is only enabled if block filters have been created using the -blockfilterindex configuration option
+  * @see [[https://bitcoincore.org/en/doc/0.19.0/rpc/blockchain/getblockfilter]]
+  */
+trait V19BlockFilterRpc {
+  self: Client =>
+
+  def getBlockFilter(
+      blockhash: String,
```
Comment (nkohen) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/V19BlockFilterRpc.scala]: This can be `DoubleSha256DigestBE`

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/V19BlockFilterRpc.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      filtertype: String): Future[GetBlockFilterResult] = {
```
Comment (nkohen) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/V19BlockFilterRpc.scala]: I believe we have a `FilterType`

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/jsonmodels/BlockchainResult.scala [hunk 3]
```diff
@@ -202,3 +202,6 @@ case class GetTxOutSetInfoResult(
     disk_size: Int,
     total_amount: Bitcoins)
     extends BlockchainResult
+
+case class GetBlockFilterResult(filter: String, header: String)
```
Comment (nkohen) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/jsonmodels/BlockchainResult.scala]: These can be `GolombFilter` and `DoubleSha256DigestBE`

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/serializers/JsonSerializers.scala [hunk 4]
```diff
@@ -229,6 +232,18 @@ object JsonSerializers {
...
 
+  implicit val setWalletFlagResultReads: Reads[SetWalletFlagResult] =
+    ((__ \ "flag_name").read[String] and
+      (__ \ "flag_state").read[Boolean] and
+      (__ \ "warnings").readNullable[String])(SetWalletFlagResult)
+
+  implicit val balanceInfoReads: Reads[BalanceInfo] =
+    ((__ \ "trusted").read[Bitcoins] and
+      (__ \ "untrusted_pending").read[Bitcoins] and
+      (__ \ "immature").read[Bitcoins])(BalanceInfo)
```
Comment (nkohen) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/serializers/JsonSerializers.scala]: I think that we should stay consistent and even use underscores in our result case classes rather than changing names and adding complexity. Really the only time where this should be used is if names are illegal like if they use the name `type` or any other reserved words

File: testkit/src/main/scala/org/bitcoins/testkit/rpc/BitcoindRpcTestUtil.scala [hunk 5]
```diff
@@ -678,6 +696,15 @@ trait BitcoindRpcTestUtil extends BitcoinSLogger {
     (BitcoindV18RpcClient, BitcoindV18RpcClient)] =
     createNodePairInternal(BitcoindVersion.V18, clientAccum)
 
+  /**
+    * Returns a pair of [[org.bitcoins.rpc.client.v19.BitcoindV19RpcClient BitcoindV19RpcClient]]
+    * that are connected with some blocks in the chain
+    */
+  def createNodePairV19(clientAccum: RpcClientAccum = Vector.newBuilder)(
```
Comment (nkohen) [inline testkit/src/main/scala/org/bitcoins/testkit/rpc/BitcoindRpcTestUtil.scala:703]: I believe you need to add a case to `createNodeSequence`.

You should get a warning for this on compile time. If it doesn't show up you can `clean` and then `compile`

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/BitcoindV19RpcClient.scala [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import play.api.libs.json.JsString
+import org.bitcoins.rpc.serializers.JsonSerializers._
+
+import scala.concurrent.Future
+import scala.util.Try
+
+/**
+  * Class for creating a BitcoindV19 instance that can access RPCs
+  * @param instance
+  * @param actorSystem
```
Comment (nkohen) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/BitcoindV19RpcClient.scala]: In general we just leave off the params unless we have something to say about them

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/BitcoindV19RpcClient.scala [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    with V19BlockFilterRpc {
+
+  override lazy val version: BitcoindVersion = BitcoindVersion.V19
+
+  /**
+    * setWalletFlag
+    *
+    * Change the state of the given wallet flag for a wallet.
+    */
+  def setWalletFlag(
```
Comment (nkohen) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/BitcoindV19RpcClient.scala:68]: nit: I don't think you need to restate the name in the scaladoc

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/BitcoindV19RpcClient.scala [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+      flag: String,
```
Comment (nkohen) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/BitcoindV19RpcClient.scala]: Is there a fixed set of `flag`s? If so we should make an ADT for this.

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/jsonmodels/WalletResult.scala [hunk 9]
```diff
@@ -50,6 +50,20 @@ case class GetTransactionResult(
     hex: Transaction)
     extends WalletResult
 
+case class SetWalletFlagResult(
+    flag: String,
```
Comment (nkohen) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/jsonmodels/WalletResult.scala]: ditto comment above on type for this

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  it should "get a block filter given a block hash" in {
+    for {
+      (client, _) <- clientPairF
+      blocks <- client.getNewAddress.flatMap(client.generateToAddress(1, _))
+      blockFilter <- client.getBlockFilter(blocks.head.hex, "basic")
+    } yield {
+      assert(blockFilter.filter.length > 0)
+      assert(blockFilter.header.length > 0)
```
Comment (nkohen) [inline bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala]: should probably assert that the block filter matches the block and the block filter header matches the block filter. Let me know if you want to know how these are computed.

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+  it should "be able to get the balances" in {
+    for {
+      (client, _) <- clientPairF
+      immatureBalance <- client.getBalances
+      _ <- client.getNewAddress.flatMap(client.generateToAddress(1, _))
+      newImmatureBalance <- client.getBalances
+    } yield {
+      assert(immatureBalance.mine.immature.toBigDecimal > 0)
+      assert(immatureBalance.mine.immature < newImmatureBalance.mine.immature)
```
Comment (nkohen) [inline bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala]: we probably know the exact difference here right?

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  it should "be able to get the balances" in {
+    for {
+      (client, _) <- clientPairF
+      immatureBalance <- client.getBalances
+      _ <- client.getNewAddress.flatMap(client.generateToAddress(1, _))
+      newImmatureBalance <- client.getBalances
+    } yield {
+      assert(immatureBalance.mine.immature.toBigDecimal > 0)
```
Comment (nkohen) [inline bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala]: Let's make this `>=` just in case this ever moves to the top of the file

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  it should "be able to set the wallet flag 'avoid_reuse'" in {
+    for {
+      (client, _) <- clientPairF
+      result <- client.setWalletFlag("avoid_reuse", value = true)
+    } yield {
+      assert(result.flag == "avoid_reuse")
+      assert(result.state)
+    }
```
Comment (nkohen) [inline bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala:75]: is there any way we can functionally check that the flag is active?

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala [hunk 14]
```diff
@@ -0,0 +1,62 @@
+package org.bitcoins.rpc.v19
+import org.bitcoins.rpc.client.common.BitcoindVersion
+import org.bitcoins.rpc.client.v19.BitcoindV19RpcClient
+import org.bitcoins.testkit.rpc.BitcoindRpcTestUtil
+import org.bitcoins.testkit.util.BitcoindRpcTest
+
+import scala.concurrent.Future
+
+class BitcoindV19RpcClientTest extends BitcoindRpcTest {
```
Comment (nkohen) [inline bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala:11]: I think we are missing an `afterAll` method here shutting things down

**Review by nkohen — COMMENTED** [review]
Looks pretty good, just a couple minor things

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/BitcoindV19RpcClient.scala (hunk 8 above)
Comment (benthecarman) [reply bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/BitcoindV19RpcClient.scala]: There is only one currently `avoid_reuse`

File: testkit/src/main/scala/org/bitcoins/testkit/rpc/BitcoindRpcTestUtil.scala (hunk 5 above)
Comment (benthecarman) [reply testkit/src/main/scala/org/bitcoins/testkit/rpc/BitcoindRpcTestUtil.scala:703]: I believe I did here https://github.com/bitcoin-s/bitcoin-s/blob/a9d896013cb6a61945f06182e95a3bffd22a3093/testkit/src/main/scala/org/bitcoins/testkit/rpc/BitcoindRpcTestUtil.scala#L619

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala (hunk 13 above)
Comment (benthecarman) [reply bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala:75]: The flag will show up in some of the updated RPCs, I could add that test case when the I update those RPCs

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala (hunk 14 above)
Comment (benthecarman) [reply bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala:11]: None of the other `RpcClientTest`s have one, should I add one to them in a future PR or do I not need one?

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/BitcoindV19RpcClient.scala (hunk 8 above)
Comment (nkohen) [reply bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/BitcoindV19RpcClient.scala]: Gotchya, well in any case I'd say it would be nice to have an ADT (Abstract Data Type) for this.

This will entail creating a
```sealed abstract class WalletFlag```
with just the one member for now, and usually for readability with things like this we put the instances inside a companion object:
```
object WalletFlag {
  case object AvoidReuse extends WalletFlag {
    override def toString: String = "avoid_reuse"
  }
}
```
Using the keyword `sealed` enforces that this `abstract class` can only be `extend`ed from within the same file. This gives us nice features like if you ever add a `WalletFlag`, you will get compiler warnings about all of the now non-exhaustive `match` statements where the new case needs to be handled.

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala (hunk 13 above)
Comment (nkohen) [reply bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala:75]: Sounds good, maybe just leave a TODO comment for now and either make yourself a note or file yourself an issue to get back to this :+1:

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala (hunk 14 above)
Comment (nkohen) [reply bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala:11]: @Christewart how is it that we don't have `afterAll`s in these tests? Am I missing something or is this definitely a problem?

Comment (Christewart) [reply bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala:11]: Take a look at the `BitcoindRpcTest` trait

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/jsonmodels/BlockchainResult.scala (hunk 3 above)
Comment (Christewart) [reply bitcoind-rpc/src/main/scala/org/bitcoins/rpc/jsonmodels/BlockchainResult.scala]: is `header` a hash or a `BlockHeader`?

File: testkit/src/main/scala/org/bitcoins/testkit/rpc/BitcoindRpcTestUtil.scala [hunk 15]
```diff
@@ -229,12 +230,14 @@ trait BitcoindRpcTestUtil extends BitcoinSLogger {
...
-      rpcUri,
-      zmqPort,
-      pruneMode,
-      blockFilterIndex = versionOpt.contains(BitcoindVersion.Experimental))
+    val configFile =
+      writtenConfig(uri,
+                    rpcUri,
+                    zmqPort,
+                    pruneMode,
+                    blockFilterIndex = versionOpt.contains(
```
Comment (Christewart) [inline testkit/src/main/scala/org/bitcoins/testkit/rpc/BitcoindRpcTestUtil.scala]: maybe move this up a few lines into a intermediate val and say 

```scala
val hasNeutrinoSupport = Vector(BitcoindVersion.Experimental, BitcoindVersion19).contains(version)
.
.
.
blockFilterIndex = hasNeutrinoSupport)
```

Comment (benthecarman) [reply testkit/src/main/scala/org/bitcoins/testkit/rpc/BitcoindRpcTestUtil.scala]: > is header a hash or a BlockHeader?

It's the header for the block filter 

https://github.com/bitcoin/bips/blob/master/bip-0157.mediawiki#filter-headers

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/jsonmodels/BlockchainResult.scala (hunk 3 above)
Comment (nkohen) [reply bitcoind-rpc/src/main/scala/org/bitcoins/rpc/jsonmodels/BlockchainResult.scala]: It is a hash since we don’t have access to additional information, specifically the previous hash

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+  it should "get a block filter given a block hash" in {
+    for {
+      (client, _) <- clientPairF
+      blocks <- client.getNewAddress.flatMap(client.generateToAddress(1, _))
+      blockFilter <- client.getBlockFilter(blocks.head, FilterType.Basic)
+
+      block <- client.getBlockRaw(blocks.head)
+      txs <- Future.sequence(
+        block.transactions
+          .filter(!_.isCoinbase)
```
Comment (nkohen) [inline bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala]: nit: `.filterNot(_.isCoinbase)`

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+          .map(x => client.getTransaction(x.txIdBE)))
```
Comment (nkohen) [inline bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala]: nit: rename `x` to `tx` or something that is a little more descriptive

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+      prevFilter <- client.getBlockFilter(block.blockHeader.previousBlockHashBE, FilterType.Basic)
+    } yield {
+      val pubKeys = txs.flatMap(_.hex.outputs.map(_.scriptPubKey)).toVector
+      assert(BlockFilter(block, pubKeys).hash == blockFilter.filter.hash)
```
Comment (nkohen) [inline bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala]: Let's make `BlockFilter(block, pubKeys)` a `val` so it isn't computed twice

File: bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+  it should "be able to get the balances" in {
+    for {
+      (client, _) <- clientPairF
+      immatureBalance <- client.getBalances
+      _ <- client.getNewAddress.flatMap(client.generateToAddress(1, _))
+      newImmatureBalance <- client.getBalances
+    } yield {
+      assert(immatureBalance.mine.immature.toBigDecimal >= 0)
+      assert(
+        immatureBalance.mine.immature.toBigDecimal + 12.5 == newImmatureBalance.mine.immature.toBigDecimal)
```
Comment (nkohen) [inline bitcoind-rpc-test/src/test/scala/org/bitcoins/rpc/v19/BitcoindV19RpcClientTest.scala]: If we could give `12.5` a name (as a `val`) that would be nice for readability

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/V19BlockFilterRpc.scala [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+trait V19BlockFilterRpc {
+  self: Client =>
+
+  private case class TempBlockFilterResult(
+      filter: String,
+      header: DoubleSha256DigestBE)
+  implicit private val tempBlockFilterResultReads: Reads[
+    TempBlockFilterResult] =
+    ((__ \ "filter").read[String] and
+      (__ \ "header").read[DoubleSha256DigestBE])(TempBlockFilterResult)
```
Comment (nkohen) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/V19BlockFilterRpc.scala]: You should be able to do a simple `Json.reads[TempBlockFilterResult]` here assuming `DoubleSha256DigestBEReads` is in scope

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/V19BlockFilterRpc.scala [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+  def getBlockFilter(
+      blockhash: DoubleSha256DigestBE,
+      filtertype: FilterType): Future[GetBlockFilterResult] = {
+    val temp = Await.result(bitcoindCall[TempBlockFilterResult](
```
Comment (nkohen) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/V19BlockFilterRpc.scala]: Rather than doing `Await.result`, this should value should be `map`ped

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/serializers/JsonSerializers.scala [hunk 22]
```diff
@@ -229,6 +229,13 @@ object JsonSerializers {
 
   implicit val bumpFeeReads: Reads[BumpFeeResult] = Json.reads[BumpFeeResult]
 
+  implicit val setWalletFlagResultReads: Reads[SetWalletFlagResult] =
+    Json.reads[SetWalletFlagResult];
+
+  implicit val balanceInfoReads: Reads[BalanceInfo] = Json.reads[BalanceInfo];
```
Comment (nkohen) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/serializers/JsonSerializers.scala]: nit: semi-colons not necessary :)

**Review by nkohen — COMMENTED** [review]
Round 2 :) Getting close!

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/V19BlockFilterRpc.scala [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+  implicit private val tempBlockFilterResultReads: Reads[
+    TempBlockFilterResult] = Json.reads[TempBlockFilterResult]
+
+  def getBlockFilter(
+      blockhash: DoubleSha256DigestBE,
+      filtertype: FilterType): Future[GetBlockFilterResult] = {
+    bitcoindCall[TempBlockFilterResult](
+      "getblockfilter",
+      List(JsString(blockhash.hex), JsString(filtertype.toString.toLowerCase)))
+      .map(t =>
```
Comment (nkohen) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/V19BlockFilterRpc.scala]: nit: more descriptive name would be nice. It is also nicer IMO to use `.map` with `{ }` but this is just an aesthetic choice

**Review by nkohen — APPROVED** [review]
LGTM, I'd like @Christewart to take another look before merging though :)

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/V19BlockFilterRpc.scala [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+  * Gets the BIP158 filter for the specified block.
+  * This RPC is only enabled if block filters have been created using the -blockfilterindex configuration option
+  * @see [[https://bitcoincore.org/en/doc/0.19.0/rpc/blockchain/getblockfilter]]
+  */
+trait V19BlockFilterRpc {
+  self: Client =>
+
+  private case class TempBlockFilterResult(
```
Comment (Christewart) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/client/v19/V19BlockFilterRpc.scala:25]: Why is this needed? Can we add a comment here indicating why we need this temp result?
